### PR TITLE
Brotli primitives

### PR DESCRIPTION
**Pull request by Vedin**
Add BrotliPrimitives and some changes to project configuration.

### CODE REVIEW

**Comment by ahsonkhan** [comment]
Are we adding the project back into the corefxlab solution now?

This PR contains a lot of old commits. Please rebase/etc as necessary.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    using nuint = System.UInt32;
+#endif 
+namespace System.IO.Compression
+{
+    public struct BrotliPrimitives
+    {
+        const int defQuality = 11;
+        const int defLgWin = 24;
+
+        public static bool Compress(ReadOnlySpan<byte> input, Span<byte> output, out nuint consumed, out nuint writen)
```
Comment (KrzysztofCwalina) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: Great to see these APIs!
Would it be possible to have the same signature as Base64 encoder @ahsonkhan just commited: 
https://github.com/dotnet/corefxlab/blob/master/src/System.Binary.Base64/System/Binary/Base64Decoder.cs#L205

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj [hunk 2]
```diff
@@ -12,24 +15,58 @@
...
+    <SourceSystem>osx</SourceSystem>
+  </PropertyGroup>
+  <PropertyGroup>
+    <BrotliPath>..\..\external\Brotli\lib\$(SourceSystem)\$(SourceArch)</BrotliPath>
+  </PropertyGroup>
+  <ItemGroup>
+    <BrotliLibraries Include="$(BrotliPath)\*.*" />
+  </ItemGroup>
+  <ItemGroup Condition=" '$(TargetsUnix)' == 'true'">
+    <Compile Include="System\IO\Compression\Interop\Library.Unix.cs" />
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: You need a version of this for OSX.

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj [hunk 3]
```diff
@@ -12,24 +15,58 @@
...
     <StartupObject />
   </PropertyGroup>
-  <!-- Platform specific properties -->
-  <PropertyGroup Condition="'$(Platform)' == 'amd64'">
+  <PropertyGroup Condition=" '$(Platform)' == 'x64'">
+    <SourceArch>x64</SourceArch>
+    <DefineConstants>BIT64;$(DefineConstants)</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Platform)' == 'x86'">
+
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: nit: remove extra space

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj [hunk 4]
```diff
@@ -12,24 +15,58 @@
...
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(Platform)' == 'arm'">
-    <PlatformTarget>arm</PlatformTarget>
-    <DefineConstants>BIT32;ARM;$(DefineConstants)</DefineConstants>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(Platform)' == 'arm64'">
-    <PlatformTarget>AnyCPU</PlatformTarget>
-    <DefineConstants>BIT64;ARM64;$(DefineConstants)</DefineConstants>
-  </PropertyGroup>
+  <Target Name="DisplayMessages">
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: Please remove this debugging information.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs [hunk 5]
```diff
@@ -69,13 +69,30 @@ public enum BrotliDecoderResult
...
+        /// 1 - For UTF-8 formatted text input
+        /// 2 - Mode used in WOFF 2.0
+        /// </summary>
+        public enum BrotliEncoderMode
+        {
+            Generic,
+            Text,
+            Font
+        };
         #region Encoder
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs]: nit: add extra spaces before and after region tag.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs [hunk 6]
```diff
@@ -109,6 +126,11 @@ public static UInt32 BrotliEncoderVersion()
         }
         #endregion
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs]: nit: add extra space around region markers.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 7]
```diff
@@ -4,7 +4,8 @@
 using System;
 using System.IO;
 using System.Runtime.InteropServices;
-
+using System.Diagnostics;
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: nit: sort using

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 8]
```diff
@@ -307,19 +308,19 @@ public override void Write(byte[] buffer, int offset, int count)
                 Marshal.Copy(buffer, currentOffset, BufferIn, copyLen);
                 bytesRemain -= copyLen;
                 currentOffset += copyLen;
-                AvailIn = (IntPtr)copyLen;
+                AvailIn = (nuint)copyLen;
                 NextIn = BufferIn;
                 while ((int)AvailIn > 0)
                 {
                     if (!BrotliNative.BrotliEncoderCompressStream(_encoder.State, BrotliNative.BrotliEncoderOperation.Process, ref AvailIn, ref NextIn, ref AvailOut,
                         ref NextOut, out totalOut)) throw new System.IO.IOException("Unable compress stream");
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Exception string need to move into a resource file.

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs [hunk 9]
```diff
@@ -19,58 +19,65 @@ internal static partial class Interop
     {
         internal static partial class Brotli
         {
-            internal const String LibName = @"C:\Users\t-detsom\Source\Repos\corefxlab\lib\brotli.dll";
-            #region Encoder
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+            internal const string LibNameEncoder = Library.BrotliEnc;
+            internal const string LibNameDecoder = Library.BrotliDec;
+#region Encoder
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: nit: spacing and tabs

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs [hunk 10]
```diff
@@ -19,58 +19,65 @@ internal static partial class Interop
...
 
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+            [DllImport(LibNameEncoder, CallingConvention = CallingConvention.Cdecl)]
             internal static extern bool BrotliEncoderIsFinished(IntPtr state);
 
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+            [DllImport(LibNameEncoder, CallingConvention = CallingConvention.Cdecl)]
             internal static extern void BrotliEncoderDestroyInstance(IntPtr state);
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+            [DllImport(LibNameEncoder, CallingConvention = CallingConvention.Cdecl)]
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: nit: add spaces between method definitions.

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs [hunk 11]
```diff
@@ -19,58 +19,65 @@ internal static partial class Interop
...
             internal static extern UInt32 BrotliEncoderVersion();
 
-            #endregion
-            #region Decoder
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+#endregion
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: nit: spacing and tabs.

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs [hunk 12]
```diff
@@ -19,58 +19,65 @@ internal static partial class Interop
...
             internal static extern void BrotliDecoderDestroyInstance(IntPtr state);
 
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+            [DllImport(LibNameDecoder, CallingConvention = CallingConvention.Cdecl)]
             internal static extern nuint BrotliDecoderVersion();
 
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
+            [DllImport(LibNameDecoder, CallingConvention = CallingConvention.Cdecl)]
             internal static extern bool BrotliDecoderIsUsed(IntPtr state);
-            [DllImport(LibName, CallingConvention = CallingConvention.Cdecl)]
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: nit: add spacing between method definitions.

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs [hunk 13]
```diff
@@ -1,10 +1,158 @@
-﻿using System;
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+using Microsoft.Xunit.Performance;
 using System.Collections.Generic;
-using System.Text;
+using System.Security.Cryptography;
```
Comment (shiftylogic) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: Sort using

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs [hunk 14]
```diff
@@ -1,10 +1,158 @@
...
+using System.Threading.Tasks;
+using System.IO.Compression;
+using System.Runtime.CompilerServices;
 
 namespace System.IO.Compression.Tests
 {
-    class BrotliPerfomanceTests
+public class BrotliPerfomanceTests
```
Comment (shiftylogic) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: nit: tabs are misaligned

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs [hunk 15]
```diff
@@ -1,10 +1,158 @@
...
     {
+        public static string GetTestFilePath(int? index = null, string memberName = null, int lineNumber = 0)
+        {
+            return Path.Combine(Path.GetTempPath(), string.Format(
+                index.HasValue ? "{0}_{1}_{2}_{3}" : "{0}_{1}_{2}",
+                memberName ?? "TestBase", lineNumber, Path.GetRandomFileName(),
+                index.GetValueOrDefault()));
+        }
+    
+    private static string CreateCompressedFile(CompressionType type)
```
Comment (shiftylogic) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: nit: spacing issue

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs [hunk 16]
```diff
@@ -1,10 +1,158 @@
...
+
+        [Benchmark]
+        [InlineData(CompressionType.CryptoRandom)]
+        [InlineData(CompressionType.RepeatedSegments)]
+        [InlineData(CompressionType.NormalData)]
+        public void Decompress(CompressionType type)
+        {
+            string testFilePath = CreateCompressedFile(type);
+
+            int _bufferSize = 1024*32;
```
Comment (shiftylogic) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: local variables should not be prefixed with underscore.

File: tests/System.IO.Compression.Tests/BrotliStreamTests.cs [hunk 17]
```diff
@@ -8,15 +8,15 @@
...
 using Xunit;
 
-
 namespace System.IO.Compression.Tests
 {
     public class BrotliStreamTests
     {
-        static string brTestFile(string fileName) => Path.Combine("..//..//..//BrotliTestData", fileName);
+        static string brTestFile(string fileName) => Path.Combine("BrotliTestData", fileName);
         [Fact]
```
Comment (shiftylogic) [inline tests/System.IO.Compression.Tests/BrotliStreamTests.cs]: nit: add an extra space

File: tests/System.IO.Compression.Tests/BrotliStreamTests.cs [hunk 18]
```diff
@@ -8,15 +8,15 @@
...
         public void BaseStreamCompress()
         {
+
```
Comment (shiftylogic) [inline tests/System.IO.Compression.Tests/BrotliStreamTests.cs]: nit: remove extra line

File: tools/common.props [hunk 19]
```diff
@@ -14,7 +14,4 @@
     <PublicSign Condition=" '$(OS)' != 'Windows_NT' ">true</PublicSign>
     <AssemblyOriginatorKeyFile>$(MSBuildThisFileDirectory)Key.snk</AssemblyOriginatorKeyFile>
   </PropertyGroup>
-  <ItemGroup>
```
Comment (shiftylogic) [inline tools/common.props]: This shouldn't be removed. This pulls in a newer Roslyn compiler that keeps our usage of Span correct. If you are having issues when using this compiler, lets discuss how to fix it.

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj (hunk 3 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: +

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj (hunk 2 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: done. Also change logic instead of include(which didn't work use remove unnecessary libraries)

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj (hunk 4 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: removed

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs (hunk 5 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs]: fix

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs (hunk 6 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliNative.cs]: +

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 7 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Is I understand correct: Sort using in  lexicographical order?

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 8 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Is it mean to create file with string constants and use it instead of "Magic strings"? May be it'll be better to create some new specialized exceptions?

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs (hunk 9 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: fix

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs (hunk 10 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: +

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs (hunk 11 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: +

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs (hunk 14 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: done

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs (hunk 15 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: fixed

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs (hunk 16 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: change to bufferSize

File: tests/System.IO.Compression.Tests/BrotliStreamTests.cs (hunk 17 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/BrotliStreamTests.cs]: +

File: tests/System.IO.Compression.Tests/BrotliStreamTests.cs (hunk 18 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/BrotliStreamTests.cs]: done and resolved the same issues in other lines.

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj [hunk 20]
```diff
@@ -4,6 +4,9 @@
     <OutputType>Library</OutputType>
   </PropertyGroup>
   <PropertyGroup>
+  <BuildPlatform>x64,x86</BuildPlatform>
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: How is this used?

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj [hunk 21]
```diff
@@ -12,24 +15,54 @@
     <OutputTypeEx>library</OutputTypeEx>
     <StartupObject />
   </PropertyGroup>
-  <!-- Platform specific properties -->
-  <PropertyGroup Condition="'$(Platform)' == 'amd64'">
+  <PropertyGroup Condition=" '$(Platform)' == 'x64'">
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: A "Chose/When" construct might be clearer:
https://github.com/dotnet/corefx/blob/master/dir.props#L133-L150

You could also add the PropertyGroup and ItemGroup under one condition.

Currently, there are duplicate checks:
```csproj
<PropertyGroup Condition=" '$(OS)' == 'Windows_NT'"> 
<ItemGroup Condition=" '$(OS)' != 'Windows_NT'"> 
```

Thanks @mellinoe for the recommendation.

@Vedin, thoughts?

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+                    bufIn = new IntPtr(inBytes);
+                    bufOut = new IntPtr(outBytes);
+                    return BrotliNative.BrotliEncoderCompress(quality, lgwin, BrotliNative.BrotliEncoderMode.Generic, consumed, bufIn, ref writen, bufOut);
+                }
+            }
+        }
+
+        public static BrotliNative.BrotliDecoderResult Decompress(ReadOnlySpan<byte> input, Span<byte> output, out nuint consumed, out nuint writen)
+        {
+            unsafe
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: Should we mark the entire method as unsafe?
```C#
public unsafe static BrotliNative.BrotliDecoderResult Decompress(ReadOnlySpan<byte> input, Span<byte> output, out nuint consumed, out nuint writen)
```

File: src/System.IO.Compression.Brotli/System/Interop/Library.OSX.cs [hunk 23]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.IO.Compression
+{
+    internal static partial class Library
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System/Interop/Library.OSX.cs]: Should we have separate files for the Library class (which get removed from the project based on the OS) that just have file name strings?

Can we somehow just set the string depending on the OS?

File: tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj [hunk 24]
```diff
@@ -18,11 +17,18 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\..\src\System.IO.Compression.Brotli\System.IO.Compression.Brotli.csproj" />
+    <ProjectReference Include="..\..\src\System.IO.Compression.Brotli\System.IO.Compression.Brotli.csproj">
+      <Private>true</Private>
```
Comment (ahsonkhan) [inline tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj]: What does `<Private>` do?

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs (hunk 12 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: fixed

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj (hunk 21 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: We use ItemGroup for different actions like(Include, Remove, Exclude, etc..) and PropertyGroup to create a set of user-defined Property elements. https://msdn.microsoft.com/en-us/library/t4w159bs.aspx 
Thank you for recommendation about Choose/When I'll think about it, because for me it isn't really clearer

Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: Something like this for example:
```csproj
  <Choose>
    <When Condition=Condition=" '$(OS)' == 'Windows_NT'"> 
      <PropertyGroup>
        <SourceSystem>windows</SourceSystem>
      </PropertyGroup>
      <ItemGroup> 
	<Compile Remove="System\Interop\Library.Unix.cs" />
	<Compile Remove="System\Interop\Library.OSX.cs" />
      </ItemGroup>
    </When>
  </Choose>
```

File: src/System.IO.Compression.Brotli/System/Interop/Library.OSX.cs (hunk 23 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/Interop/Library.OSX.cs]: I think we can with using define constants, but to be convenient I did it like in corefx https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Windows/Interop.Libraries.cs (and the same for unix and osx)

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj (hunk 21 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: In this case we should duplicate some removes, like write:
`<Compile Remove="System\Interop\Library.Unix.cs" />` in windows When and in OSX When both

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 8 above)
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Regardless of having a custom exception, strings should be in a resource files. Take a look at how exceptions are thrown [here](https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression/src/System/IO/Compression/GZipStream.cs)

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj [hunk 25]
```diff
@@ -12,24 +15,54 @@
...
+  </ItemGroup>
+  <ItemGroup Condition=" '$(OS)' != 'OSX'">
+    <Compile Remove="System\Interop\Library.OSX.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Content Include="$(BrotliPath)\**">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+  <PropertyGroup Condition="'$(Platform)' == 'x64'">
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: Combine this PropertyGroup with the one above.

File: tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj (hunk 24 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj]: Specifies whether the reference should be copied to the output folder.

File: tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs (hunk 13 above)
Comment (Vedin) [inline tests/System.IO.Compression.Tests/BrotliPerfomanceTests.cs]: ok

File: src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs (hunk 12 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/Interop/Interop.Brotli.cs]: +

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs (hunk 1 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: Changed

File: src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj (hunk 25 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System.IO.Compression.Brotli.csproj]: done

File: tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj (hunk 24 above)
Comment (ahsonkhan) [inline tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj]: Do we want it to be copied to the output folder? Why is this necessary?

Comment (Vedin) [inline tests/System.IO.Compression.Tests/System.IO.Compression.Tests.csproj]: Because it won' work without it.

File: tools/common.props (hunk 19 above)
Comment (Vedin) [inline tools/common.props]: I don't remember why I delete it, but I try and now everything is ok.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 8 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: done in other way, without using a SR.cs

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs (hunk 22 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: IMHO:  not necessary

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+#endif 
+
+namespace System.IO.Compression
+{
+    public struct BrotliPrimitives
+    {
+        const int defQuality = 11;
+        const int defLgWin = 24;
+
+        public static bool Compress(ReadOnlySpan<byte> source, Span<byte> destination, out nuint bytesConsumed, out nuint bytesWritten)
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: Do bytesConsumed/bytesWritten need to be nuint or can they just be ints?

Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: Yes, they should be nuint, because 1. array size can't be negative(so number of bytes in byte array also). 2. the native signature is: `size_t input_size,
    const uint8_t* input_buffer, size_t* encoded_size,
    uint8_t* encoded_buffer` ans size_t is unsigned type.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 27]
```diff
@@ -42,7 +46,7 @@ public partial class BrotliStream : Stream
         }
         public BrotliStream(Stream baseStream, CompressionMode mode, bool leaveOpen, int BuffSize)
         {
-            if (baseStream == null) throw new ArgumentNullException("baseStream");
+            if (baseStream == null) throw new ArgumentNullException(BrotliEx.baseStream);
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: These types of strings don't need to be in resources... because they need to match the parameter names. They should stay in code.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 28]
```diff
@@ -185,19 +190,19 @@ private void ValidateParameters(byte[] array, int offset, int count)
                 throw new ArgumentOutOfRangeException(nameof(count));
 
             if (array.Length - offset < count)
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: The spacing looks off here.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 29]
```diff
@@ -33,7 +37,7 @@ public partial class BrotliStream : Stream
         Decoder _decoder;
         Encoder _encoder;
         public BrotliStream(Stream baseStream, CompressionMode mode, bool leaveOpen, int BuffSize, uint windowSize, uint quality) : this(baseStream, mode, leaveOpen, BuffSize) {
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Please format this code with open brace on new line and possibly newline + indent for calling the other constructor.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 30]
```diff
@@ -33,7 +37,7 @@ public partial class BrotliStream : Stream
...
-            if (_mode == CompressionMode.Decompress) throw new System.IO.IOException("quality and windowsize is ambitious for Decompress mode");
+            if (_mode == CompressionMode.Decompress) throw new System.IO.IOException(BrotliEx.QualityAndWinSize);
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Format code:

if (condition)
    throw ...

// No else necessary here.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs [hunk 31]
```diff
@@ -185,19 +190,19 @@ private void ValidateParameters(byte[] array, int offset, int count)
                 throw new ArgumentOutOfRangeException(nameof(count));
 
             if (array.Length - offset < count)
-                throw new ArgumentException("InvalidArgument","OffsetCount");
+                throw new ArgumentException(BrotliEx.InvalidArgument,BrotliEx.OffsetCount);
```
Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Throw ArgumentOutOfRange exception here instead.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 29 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: done

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 30 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: ok

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 27 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Is it only for Agrument Exceptions?

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs [hunk 32]
```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+#if BIT64
+    using nuint = System.UInt64;
+#else
+using nuint = System.UInt32;
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: nit: indentation inconsistency.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs (hunk 28 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliStream.cs]: Do you mean delete new lines between if statements?

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs (hunk 32 above)
Comment (Vedin) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: Yes, it is interesting. I just have done Ctrl-A, Ctrl-K, Ctrl+F and it changed, but it doesn't look like standard style.

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            return Compress(source, destination, out bytesConsumed, out bytesWritten, defQuality, defLgWin);
+        }
+
+        public static bool Compress(ReadOnlySpan<byte> source, Span<byte> destination, out nuint bytesConsumed, out nuint bytesWritten, int quality, int lgwin)
+        {
+            unsafe
+            {
+                bytesConsumed = (nuint)source.Length;
+                bytesWritten = (nuint)destination.Length;
```
Comment (ahsonkhan) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: bytesWritten should start as 0.

File: tests/System.Text.Primitives.Tests/Brotli64.xml [hunk 34]
```diff
@@ -0,0 +1,3072 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (ahsonkhan) [inline tests/System.Text.Primitives.Tests/Brotli64.xml]: Do we want this file checked in?

Comment (Vedin) [inline tests/System.Text.Primitives.Tests/Brotli64.xml]: No

File: src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs (hunk 26 above)
Comment (whoisj) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: @Vedin that's not a great argument in .NET. While we all agree that size cannot be less than zero, not all .NET languages support unsigned integers. Besides, I believe @ahsonkhan is asking if the size needs to be `size_t` or if the signature can just be fixed.

Comment (shiftylogic) [inline src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs]: I agree these should be int at the public layer. The reasoning is to match similar APIs (base64, encoders, etc) but also because Span only supports int for indexer and thus only allows access to int.MaxValue units in the Span.

I think we should merge this PR. We have another change coming that is going to change the signatures of the public surface to follow a new convention similar to our other Span APIs.
